### DISCUSSION TITLE
How to bundle KaTeX CSS to a single output CSS with webpack?

### DISCUSSION THREAD

**Discussion by cirosantilli** [KaTeX/KaTeX#3115] [Q&A] [answered]
I would like to produce a single bundled CSS file. In the example below, normalize works if I comment out the KaTeX part, so I know I'm on the right track, but I haven't managed to get the KaTeX part to work.

First I tried:

main.scss

```
@use "normalize.css/normalize.css";
@use "katex/dist/katex.min.css";

body {
  background-color: red;
}
```

webpack.config.js

```
const path = require('path');

const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

const nodeModulesPath = path.resolve(__dirname, 'node_modules');

module.exports = {
  entry: {
    main: ['./main.scss'],
  },
  mode: 'none',
  module: {
    rules: [
      {
        test: /\.(scss|css)$/,
        use: [
          {
            loader: MiniCssExtractPlugin.loader,
          },
          {
            loader:  'css-loader',
            options: {
              sourceMap: true,
            },
          },
          // 'resolve-url-loader',
          {
            loader: "sass-loader",
            options: {
              sassOptions: {
                includePaths: [nodeModulesPath],
              },
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: '[name].css',
    }),
  ],
  optimization: {
    minimizer: [
      new CssMinimizerPlugin(),
    ],
    minimize: true,
  },
};
```

package.json

```
{
  "name": "webpack-cheat",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "build": "webpack",
    "sass": "rm -f dist/main.css && sass main.scss dist/main.css"
  },
  "author": "Ciro Santilli",
  "license": "MIT",
  "devDependencies": {
    "css-loader": "5.2.4",
    "css-minimizer-webpack-plugin": "3.0.2",
    "katex": "^0.13.11",
    "mini-css-extract-plugin": "2.1.0",
    "normalize.css": "8.0.1",
    "resolve-url-loader": "^4.0.0",
    "sass": "1.32.11",
    "sass-loader": "11.0.1",
    "style-loader": "2.0.0",
    "webpack": "5.36.1",
    "webpack-cli": "4.6.0",
    "webpack-dev-server": "3.11.2"
  }
}
```

which fails with:

```
Error: Can't resolve 'fonts/KaTeX_AMS-Regular.woff2' in '/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import'
```

presumably because `node_modules/katex/dist/katex.css` does:

```
@font-face {
  font-family: 'KaTeX_AMS';
  src: url(fonts/KaTeX_AMS-Regular.woff2) format('woff2'), url(fonts/KaTeX_AMS-Regular.woff) format('woff'), url(fonts/KaTeX_AMS-Regular.ttf) format('truetype');
```

and `fonts/KaTeX_AMS-Regular.ttf` is at `node_modules/katex/dist/`fonts/KaTeX_AMS-Regular.ttf`, and relative paths are not being searched.

 This was asked at: https://stackoverflow.com/questions/54042335/webpack-and-fonts-with-relative-paths so then I tried to use  `resolve-url-loader'` by uncommenting the line `// 'resolve-url-loader' above, but then it fails with:

```
ERROR in ./main.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/resolve-url-loader/index.js):
Error: resolve-url-loader: error processing CSS
  a valid source-map is not present (ensure preceding loaders output a source-map)
  at file:///home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/main.scss:344:3
    at encodeError (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/resolve-url-loader/index.js:287:12)
    at onFailure (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/resolve-url-loader/index.js:228:14)
    at processResult (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/webpack/lib/NormalModule.js:676:19)
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/webpack/lib/NormalModule.js:778:5
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/loader-runner/lib/LoaderRunner.js:399:11
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/loader-runner/lib/LoaderRunner.js:251:18
    at context.callback (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/loader-runner/lib/LoaderRunner.js:124:13)
    at onFailure (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/resolve-url-loader/index.js:228:5)
```

asking at: https://github.com/bholloway/resolve-url-loader/issues/212

Any clues? Also asked at: https://stackoverflow.com/questions/68366936/how-to-bundle-katex-css-to-a-single-output-css-with-webpack

**Comment by edemaine** [comment]
Any chance you can use `node_modules/katex/src/katex.less`? That points to the correct font directory (`../fonts` instead of `fonts/`).

**Reply by cirosantilli** [reply]
Thanks for this suggestion. Do you know if it is possible to use the .less from the .scss somehow?

The main.scss would be the websites main style, so I'd rather not change from Sass to Less for this reason.

But Webpack is so magic that maybe there's a way, even if it involves an intermediate CSS stage.

I tried to quickly hack `npm install less less-loader --save-dev` and:

wepack.config.js 

```
const path = require('path');

const CssMinimizerPlugin = require('css-minimizer-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');

const nodeModulesPath = path.resolve(__dirname, 'node_modules');

module.exports = {
  entry: {
    main: ['./main.scss'],
  },
  mode: 'none',
  module: {
    rules: [
      {
        test: /\.(scss|css)$/,
        use: [
          MiniCssExtractPlugin.loader,
          'css-loader',
          {
            loader: "sass-loader",
            options: {
              sassOptions: {
                includePaths: [nodeModulesPath],
              },
            },
          },
        ],
      },
      {
        test: /\.(less)$/,
        use: [
          MiniCssExtractPlugin.loader,
          'css-loader',
          {
            loader: 'less-loader',
            options: {
              lessOptions: {
                paths: [nodeModulesPath],
              },
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: '[name].css',
    }),
  ],
  optimization: {
    minimizer: [
      new CssMinimizerPlugin(),
    ],
    minimize: true,
  },
};
```

but it says it can't find the file:

```
ERROR in ./main.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
SassError: Can't find stylesheet to import.
  ╷
2 │ @use "katex/src/katex.less";
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
  main.scss 2:1  root stylesheet
    at processResult (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/webpack/lib/NormalModule.js:676:19)
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/webpack/lib/NormalModule.js:778:5
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/loader-runner/lib/LoaderRunner.js:399:11
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/loader-runner/lib/LoaderRunner.js:251:18
    at context.callback (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/loader-runner/lib/LoaderRunner.js:124:13)
    at /home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/sass-loader/dist/index.js:54:7
    at Function.call$2 (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/sass/sass.dart.js:92571:16)
    at _render_closure1.call$2 (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/sass/sass.dart.js:81074:12)
    at _RootZone.runBinary$3$3 (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/sass/sass.dart.js:27256:18)
    at _FutureListener.handleError$1 (/home/ciro/bak/git/cirosantilli.github.io/web-cheat/webpack/sass-import/node_modules/sass/sass.dart.js:25812:19)
```

**Reply by edemaine** [reply]
Can you make a `main.less` file with the line `@import "katex/src/katex.less"`, in addition to your `main.scss` file (which no longer `@use`s katex)?

**Reply by cirosantilli** [reply]
Found some time to try it out. Here is the current summary of results:

- `@use` from Sass file: couldn't get it to work: https://github.com/cirosantilli/cirosantilli.github.io/blob/284903bbc5c5e22993d69182d61d62bd55757e50/web-cheat/webpack/sass/main.scss#L5
- `@import` from Less file: worked with the .css, not the .less file: https://github.com/cirosantilli/cirosantilli.github.io/blob/284903bbc5c5e22993d69182d61d62bd55757e50/web-cheat/webpack/less/main.less#L5
- `@import` from the .js entry: worked. https://github.com/cirosantilli/cirosantilli.github.io/blob/284903bbc5c5e22993d69182d61d62bd55757e50/web-cheat/webpack/template/index.js#L8

  This is perhaps the most common use case.

  EDIT: ignore this paragraph: The reason I embarked on the alternate "separate `.css`" attempt was to try and get the .css used automatically as a transitive dependency from another module by embedding it. If I use the .js approach, I also have to re-include the CSSes on the dependant as seen here: https://github.com/cirosantilli/cirodown/blob/master/web/pages/_app.tsx#L12 which includes both the katex and normalize, which are actually dependencies of the toplevel cirodown package: https://github.com/cirosantilli/cirodown/blob/59968b0a61218ff0a2cf325d7811a9b59aa4faa4/main.scss

One thing to note, a key ingredient of getting things to work was:

```
  module: {
    rules: [
      {
      // Fonts.
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/i,
        type: 'asset/resource',
      },
```

**Reply by edemaine** [reply]
Thanks for your experiments! I'm surprised that the Less `@import` didn't work, but maybe the Webpack Less library doesn't rewrite paths as it should...

It might be nice to add some documentation about using KaTeX with Webpack. Perhaps adding to https://katex.org/docs/libs.html? Would you be willing to write a PR about your successful approach (`import` from JS), along with the font rule?

By the way, I did open #3116 to discuss switching to SCSS which might simplify things. But we can still document how things are now, and update after such a switch.

I'm not sure I understand the dependency issue... Perhaps the dependent should allow the parent to say "I'll include KaTeX's CSS; please don't do it yourself" (like a peer dependency)?

**Reply by cirosantilli** [reply]
> Thanks for your experiments! I'm surprised that the Less `@import` didn't work, but maybe the Webpack Less library doesn't rewrite paths as it should...
> 
> It might be nice to add some documentation about using KaTeX with Webpack. Perhaps adding to https://katex.org/docs/libs.html? Would you be willing to write a PR about your successful approach (`import` from JS), along with the font rule?
> 

Yes, it would be good to document this, especially with a minimal demo in a repo under the KaTeX org. EDIT: better, in a directory under https://github.com/KaTeX/KaTeX/tree/master/docs each of those setups is so small, better keep in-tree.

I would then put a link to those repos under the browser installation page: https://katex.org/docs/browser.html

Before making this, I just wish we would understand more clearly why some things work and others don't.

> By the way, I did open #3116 to discuss switching to SCSS which might simplify things. But we can still document how things are now, and update after such a switch.
> 

Good to hear about that.

> I'm not sure I understand the dependency issue... Perhaps the dependent should allow the parent to say "I'll include KaTeX's CSS; please don't do it yourself" (like a peer dependency)?

I thought about this a bit more, import from Js would likely work for transitive dependencies, I was confused with other complications of my setup I think.

**Reply by edemaine** [reply]
I found [this fixed bug](https://github.com/webpack-contrib/less-loader/issues/76) which might have been the issue, fixed in 2020. Are you using the latest less-loader (10.0.1)?

**Reply by cirosantilli** [reply]
> I found [this fixed bug](https://github.com/webpack-contrib/less-loader/issues/76) which might have been the issue, fixed in 2020. Are you using the latest less-loader (10.0.1)?

Yes, I'm on that version: https://github.com/cirosantilli/cirosantilli.github.io/blob/284903bbc5c5e22993d69182d61d62bd55757e50/web-cheat/webpack/less/package.json#L15

**Reply by edemaine** [reply]
Ah, I see now why Less `@import "katex/src/katex.less"` doesn't work. [That file](https://github.com/KaTeX/KaTeX/blob/25a81c67f68c1c536f3a7b5bdee62278328a48fc/src/katex.less#L2) does `@import "fonts.less"`, and [fonts.less](https://github.com/KaTeX/KaTeX/blob/25a81c67f68c1c536f3a7b5bdee62278328a48fc/src/fonts.less#L1) uses the path `"../fonts"`. So far so good -- the repo has a root `fonts` directory.  However, in the NPM distribution, it's in `dist`.  Hence the problem.

One possible fix here would be to distribute a version of `katex.less` (possibly even with an overridable fonts directory variable) within `dist` in NPM. Then you could use that version instead of the one in `src`. But given that the CSS version works, I think that's the best thing to recommend, until #3116 happens.

**Reply by HenrikStaaf** [reply]
@cirosantilli What is the reason for wanting to produce a single bundled CSS file? Does it help with page speeds?

**Reply by cirosantilli** [reply]
@HenrikStaaf the main issue for me was not performance, but transitive dependencies/sanity of distribution.

E.g. https://github.com/cirosantilli/cirodown is a markup language with KaTeX support out of box.

That package uses KaTeX and CSS from various other sources.

Therefore, the sanest way to distribute the CSS for that markup language is to package *everything* into a single CSS file, that way end users only need to include a single `cirodown.css` file to make output render correctly.

And one of the sanest ways to package a bunch of CSS files into one is to run webpack before any releases of the markup and add the .css to the NPM package.

Now I'm being forced to do some hacks calling sass directly without webpack: https://github.com/cirosantilli/cirodown/issues/157

**Reply by HenrikStaaf** [reply]
@cirosantilli I see, I am currently looking for a fix to the huge DOM trees that Katex produces. I love Katex but it is killing the page speed of my math pages :(. I thought this may be one way to fix these performance issues.

**Reply by cirosantilli** [reply]
@HenrikStaaf ah I see. Yeah, I suppose this is a hard problem, besides the more obvious options of either going more native (will browsers every support mathml??) or going more reactive and do it on Js after HTML page render (and thus become mathjax slow load horror). Out of curiosity, is this for Mathleaks  BTW/are you https://www.linkedin.com/in/henrik-s-a15b6b67/ ? Interesting project.

**Comment by HenrikStaaf** [comment]
Yes correct, Im doing this for Mathleaks :)

**Comment by cirosantilli** [comment, marked as answer]
The amazing maintainer of resolve-url-loader pointed me out the solution: https://github.com/bholloway/resolve-url-loader/issues/212#issuecomment-1011630220 needed to add `{sourceMap: true}` working example at: https://github.com/cirosantilli/cirosantilli.github.io/tree/c4a93c23e9e009a0a36e76673d167c6149295dec/web-cheat/webpack/sass cc. @HenrikStaaf  @edemaine

**Reply by HenrikStaaf** [reply]
Hi @cirosantilli.

What was solved here? I am sorry for not understanding :) is this a fix to the DOM tree issue?

**Reply by cirosantilli** [reply]
Ah, not to the DOM thing, only to the katex + webpack without manually loading fonts.

**Reply by HenrikStaaf** [reply]
Does this increase loading speeds?

**Reply by cirosantilli** [reply]
I don't think it will. My main concern was to not duplicate those includes across the package dependency chain.

**Reply by HenrikStaaf** [reply]
i understand. Thank you for the update :)
